### PR TITLE
Web Inspector: Getting style data via WebCore::InspectorCSSAgent::getMatchedStylesForNode for a detached node crashes

### PR DESCRIPTION
**Pull request by patrickangle**
#### 4c354a6ba2286680e337d91f3a09ec60ee5bf5bf
<pre>
Web Inspector: Getting style data via WebCore::InspectorCSSAgent::getMatchedStylesForNode for a detached node crashes
<a href="https://bugs.webkit.org/show_bug.cgi?id=240608">https://bugs.webkit.org/show_bug.cgi?id=240608</a>
rdar://93473674

Reviewed by Devin Rousso.

Added test case to inspector/css/node-styles-refreshed.html

Previously it was possible under at least two different circumstances that
`WebCore::InspectorCSSAgent::getMatchedStylesForNode` could be called with the NodeId of a detached node, which means
that `computedStyle` for that Node will be `nullptr`, and calls to `Style::Scope::forNode` will fail.

The first failure state is timing based. If the Inspector frontend makes a request for updated styles for a node, the
node may already have been detached by some other means (webpage JS, for example). In this case by the time
`getMatchedStylesForNode` is invoked the `computedStyle` is no longer available.

The second failure state is easier to get into. In a DOM tree with elements A, B, and C, each a child of the
previous such that A is the parent of B and B is the parent of C, imagine C is the selected DOM node. Now if
we right click and delete A from the DOM tree, we will have removed all three elements from the tree (both the
frontend representation as well as the actual tree in the backend). Previously C remained the selected node in the
tree, which means that we continue to show the Styles sidebar panel, where the user could then attempt to edit the style
of the node. These style changes trigger the need to refresh the frontend&apos;s style information, at which point we will
call `getMatchedStylesForNode` with a detached node id.

The fix to the underlying problem is to check in the backend that a node is actually attached before determining its
style information. Additionally, we should always update the selection of a TreeOutline when the current selection is
part of the subtree being removed from the tree. This prevents the user being able to easily get into the state in the
first place as well as prevents folks from getting confused about what the open sidebars are showing information for,
since visually no node _appears_ selected.

* Source/WebCore/inspector/agents/InspectorCSSAgent.cpp:
(WebCore::InspectorCSSAgent::getMatchedStylesForNode):
(WebCore::InspectorCSSAgent::getComputedStyleForNode):
- Guard the functions that will eventually use `computedStyle` or `Style::Scope::forNode`, both of which only work for
&quot;connected&quot; (in Web Inspector parlance &quot;attached&quot;) nodes. Note there are a few other uses in our code where
`computedStyle` is used without this check, but we do explicitly check the result of computed styles in those cases,
which also works since `computedStyle` returns early if the node is not connected.

(WebCore::InspectorCSSAgent::buildObjectForRule):
- Add an assertion that we never call buildObjectForRule with a disconnected Element.

* Source/WebInspectorUI/UserInterface/Views/TreeOutline.js:
(WI.TreeOutline.prototype.removeChildAtIndex):
- Fix selection updating to account for the fact the selected node may be part of the descendants tree of the removed
item, in which case we would still want to update the selection to something else.

* LayoutTests/inspector/css/node-styles-refreshed-expected.txt:
* LayoutTests/inspector/css/node-styles-refreshed.html:
- Add test case to try getting styles of a detached node, and update other test cases with a new helper function instead
of relying on a &quot;global&quot; `nodeStyles`.

Canonical link: <a href="https://commits.webkit.org/250769@main">https://commits.webkit.org/250769@main</a>
git-svn-id: <a href="https://svn.webkit.org/repository/webkit/trunk@294512">https://svn.webkit.org/repository/webkit/trunk@294512</a> 268f45cc-cd09-0410-ab3c-d52691b4dbfc
</pre>

### CODE REVIEW

File: Source/WebCore/inspector/agents/InspectorCSSAgent.cpp [hunk 1]
```diff
@@ -585,13 +591,13 @@ static Ref<Protocol::CSS::Font> buildObjectForFont(const Font& font)
             .setMaximumValue(variationAxis.maximumValue())
             .setDefaultValue(variationAxis.defaultValue())
             .release();
-        
+
```
Comment (dcrousso) [inline Source/WebCore/inspector/agents/InspectorCSSAgent.cpp]: NIT: while I do personally like the fact that we're fixing unnecessary whitespace, it's probably best to leave as-is and fix when changing code in this area

File: Source/WebCore/inspector/agents/InspectorCSSAgent.cpp [hunk 2]
```diff
@@ -491,6 +491,9 @@ Protocol::ErrorStringOr<std::tuple<RefPtr<JSON::ArrayOf<Protocol::CSS::RuleMatch
     if (!element)
         return makeUnexpected(errorString);
 
+    if (!element->isConnected())
```
Comment (dcrousso) [inline Source/WebCore/inspector/agents/InspectorCSSAgent.cpp:494]: NIT: I wish we had something like `Element::canGetStyle` or `Element::hasComputedStyle` so that we can unify all these `isConnected` that are actually related.  I suppose that's what `isConnected` does, but I'd hate for it to change in the future and this (or something else) to get mistakenly left behind.

File: Source/WebCore/inspector/agents/InspectorCSSAgent.cpp [hunk 3]
```diff
@@ -1099,6 +1105,8 @@ RefPtr<Protocol::CSS::CSSRule> InspectorCSSAgent::buildObjectForRule(const Style
     if (!styleRule)
         return nullptr;
 
+    ASSERT(element.isConnected());
```
Comment (dcrousso) [inline Source/WebCore/inspector/agents/InspectorCSSAgent.cpp]: NIT: Should this be at the top?  Or is there a scenario where `styleRule` is `nullptr` and the `element` is not attached?

File: Source/WebInspectorUI/UserInterface/Views/TreeOutline.js [hunk 4]
```diff
@@ -316,13 +316,19 @@ WI.TreeOutline = class TreeOutline extends WI.Object
         let child = this.children[childIndex];
         let parent = child.parent;
 
-        if (child.deselect(suppressOnDeselect) && !suppressSelectSibling) {
-            if (child.previousSibling)
-                child.previousSibling.select(true, false);
-            else if (child.nextSibling)
-                child.nextSibling.select(true, false);
+        let currentSelectedElement = this.treeOutline.selectedTreeElement;
+        let childOrDescentdantWasSelected = !!child.selfOrDescendant((descentdant) => descentdant === currentSelectedElement);
```
Comment (dcrousso) [inline Source/WebInspectorUI/UserInterface/Views/TreeOutline.js]: s/`childOrDescentdantWasSelected`/`childOrDescendantWasSelected`
s/`descentdant`/`descendant`

NIT: the `!!` is kinda unnecessary since you're not using `childOrDescendantWasSelected` for anything other than a truthy check

File: Source/WebInspectorUI/UserInterface/Views/TreeOutline.js [hunk 5]
```diff
@@ -316,13 +316,19 @@ WI.TreeOutline = class TreeOutline extends WI.Object
         let child = this.children[childIndex];
         let parent = child.parent;
 
-        if (child.deselect(suppressOnDeselect) && !suppressSelectSibling) {
-            if (child.previousSibling)
-                child.previousSibling.select(true, false);
-            else if (child.nextSibling)
-                child.nextSibling.select(true, false);
+        let currentSelectedElement = this.treeOutline.selectedTreeElement;
```
Comment (dcrousso) [inline Source/WebInspectorUI/UserInterface/Views/TreeOutline.js]: How does this work when multiple things are selected?  Maybe we should just check `descendant.selected` instead?

File: Source/WebInspectorUI/UserInterface/Views/TreeOutline.js [hunk 6]
```diff
@@ -316,13 +316,19 @@ WI.TreeOutline = class TreeOutline extends WI.Object
...
+        let childOrDescentdantWasSelected = !!child.selfOrDescendant((descentdant) => descentdant === currentSelectedElement);
+
+        let childWasSelected = child.deselect(suppressOnDeselect);
+
+        if (childOrDescentdantWasSelected && !suppressSelectSibling) {
+            const omitFocus = true;
+            if (childWasSelected && child.previousSibling)
```
Comment (dcrousso) [inline Source/WebInspectorUI/UserInterface/Views/TreeOutline.js]: I kinda feel like we should still select the `child.previousSibling`/`child.nextSibling` even if the `child` itself is not selected, as the `child` is still the "lowest common ancestor" of the selection, so we'd probably want the new selection to be as close to the old selection as possible.

File: LayoutTests/inspector/css/node-styles-refreshed.html [hunk 7]
```diff
@@ -55,26 +79,39 @@
...
+        name: "NodeStylesRefreshed.DisconnectedNode",
+        description: "Ensure that changing the style of a disconnected node does not crash.",
+        async test() {
+            let nodeStyles = await nodeStylesForSelector("#removable-node");
+
+            nodeStyles.singleFireEventListener(WI.DOMNodeStyles.Event.Refreshed, (event) => {
+                InspectorTest.expectFalse(event.data.significantChange, `Refreshing styles of an unchanged node should not be a significant change.`);
+            });
+
+            InspectorTest.log("-Refreshing styles before removal.")
```
Comment (dcrousso) [inline LayoutTests/inspector/css/node-styles-refreshed.html]: NIT: IMO having a line start with `-` makes it look like someone forgot to remove something from a diff/merge 😅

**Review by dcrousso — APPROVED** [review]
Nice work!

> The second failure state is easier to get into. In a DOM tree with elements \<A >, \<B >, and \<C >, each a child of the

I think the space before the `>` is getting added cause otherwise it renders as actual HTML.  Maybe just call them `A`, `B`, and `C` instead?

> frontend represetnation as well as the actual tree in the backend). Previosuly <C > remained the selected node in the

s/represetnation/representation
s/Previosuly/Previously

File: Source/WebCore/inspector/agents/InspectorCSSAgent.cpp (hunk 3 above)
Comment (patrickangle) [reply Source/WebCore/inspector/agents/InspectorCSSAgent.cpp]: Yeah, it should probably be at the top to match our general style.
